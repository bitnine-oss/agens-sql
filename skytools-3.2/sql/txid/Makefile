
PG_CONFIG = pg_config
PGXS = $(shell $(PG_CONFIG) --pgxs)

PGVER := $(shell $(PG_CONFIG) --version | sed 's/PostgreSQL //')

ifeq ($(PGVER),)
$(error Failed to get Postgres version)
else
# postgres >= manages epoch itself, so skip epoch tables
pg83 = $(shell test $(PGVER) "<" "8.3" && echo "false" || echo "true")
pg82 = $(shell test $(PGVER) "<" "8.2" && echo "false" || echo "true")
endif

ifeq ($(pg83),true)  # we have 8.3 with internal txid

# install empty txid.sql
DATA_built = txid.sql
include $(PGXS)
txid.sql: txid.internal.sql
	cp $< $@

else # 8.2 or 8.1
#
# pg < 8.3 needs this module
#
MODULE_big = txid
SRCS = txid.c epoch.c
OBJS = $(SRCS:.c=.o)
REGRESS = txid
REGRESS_OPTS = --load-language=plpgsql
DATA = uninstall_txid.sql
DOCS = README.txid
DATA_built = txid.sql
EXTRA_CLEAN = txid.sql.in

# PGXS build procedure
include $(PGXS)

ifeq ($(pg82),true)
# 8.2 tracks epoch internally
TXID_SQL = txid.std.sql
else
# 8.1 needs epoch-tracking code
TXID_SQL = txid.std.sql txid.schema.sql
endif # ! 8.2

# additional deps
txid.o: txid.h
epoch.o: txid.h

txid.sql.in: $(TXID_SQL)
	cat $(TXID_SQL) > $@

endif # ! 8.3

test: install
	make installcheck || { less regression.diffs; exit 1; }

