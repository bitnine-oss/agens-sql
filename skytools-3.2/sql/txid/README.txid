
txid - 8 byte transaction ID's
==============================

Based on xxid module from Slony-I.  

The goal is to make PostgreSQL internal transaction ID and snapshot
data usable externally.  They cannot be used directly as the
internal 4-byte value wraps around and thus breaks indexing.

This module extends the internal value with wraparound cound (epoch).
It uses relaxed method for wraparound check.  There is a table
txid.epoch (epoch, last_value) which is used to check if the xid
is in current, next or previous epoch.  It requires only occasional
read-write access - ca. after 100k - 500k transactions.

Also it contains type 'txid_snapshot' and following functions:


txid_current() returns int8

  Current transaction ID

txid_current_snapshot() returns txid_snapshot

  Current snapshot

txid_snapshot_xmin( snap ) returns int8

  Smallest TXID in snapshot.  TXID's smaller than this
  are all visible in snapshot.

txid_snapshot_xmax( snap ) returns int8

  Largest TXID in snapshot.  TXID's starting from this one are
  all invisible in snapshot.
	
txid_snapshot_xip( snap ) setof int8

  List of uncommitted TXID's in snapshot, that are invisible
  in snapshot.  Values are between xmin and xmax.

txid_visible_in_snapshot(id, snap) returns bool

  Is TXID visible in snapshot?


Problems
--------

- it breaks when there are more than 2G tx'es between calls.
  Fixed in 8.2
 
- functions that create new txid's should be 'security definers'
  thus better protecting txid_epoch table.

- After loading database from backup you should do:

  UPDATE txid.epoch SET epoch = epoch + 1,
                        last_value = (get_current_txid() & 4294967295);

