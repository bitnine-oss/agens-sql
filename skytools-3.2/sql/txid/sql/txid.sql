-- init
\set ECHO none
\i txid.sql
\set ECHO all

-- i/o
select '12:13:'::txid_snapshot;
select '12:13:1,2'::txid_snapshot;

-- errors
select '31:12:'::txid_snapshot;
select '0:1:'::txid_snapshot;
select '12:13:0'::txid_snapshot;
select '12:13:2,1'::txid_snapshot;

create table snapshot_test (
	nr	integer,
	snap	txid_snapshot
);

insert into snapshot_test values (1, '12:13:');
insert into snapshot_test values (2, '12:20:13,15,18');
insert into snapshot_test values (3, '100001:100009:100005,100007,100008');

select snap from snapshot_test order by nr;

select  txid_snapshot_xmin(snap),
	txid_snapshot_xmax(snap),
	txid_snapshot_xip(snap)
from snapshot_test order by nr;

select id, txid_visible_in_snapshot(id, snap)
from snapshot_test, generate_series(11, 21) id
where nr = 2;

-- test current values also
select txid_current() >= txid_snapshot_xmin(txid_current_snapshot());
-- select txid_current_txid() < txid_snapshot_xmax(txid_current_snapshot());

-- select txid_in_snapshot(txid_current_txid(), txid_current_snapshot()),
--    txid_not_in_snapshot(txid_current_txid(), txid_current_snapshot());

