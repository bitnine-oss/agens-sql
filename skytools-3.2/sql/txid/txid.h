#ifndef _TXID_H_
#define _TXID_H_

#define MAX_INT64  0x7FFFFFFFFFFFFFFFLL

/* Use unsigned variant internally */
typedef uint64 txid;

typedef struct
{
    int32       __varsz;   /* should not be touched directly */
    uint32      nxip;
    txid xmin;
    txid xmax;
    txid xip[1];
}   TxidSnapshot;

#define TXID_SNAPSHOT_SIZE(nxip) (offsetof(TxidSnapshot, xip) + sizeof(txid) * nxip)

typedef struct {
	uint64		last_value;
	uint64		epoch;
}	TxidEpoch;

/* internal functions */
void	txid_load_epoch(TxidEpoch *state, int try_write);
txid	txid_convert_xid(TransactionId xid, TxidEpoch *state);

/* public functions */
Datum       txid_current(PG_FUNCTION_ARGS);
Datum       txid_current_snapshot(PG_FUNCTION_ARGS);

Datum       txid_snapshot_in(PG_FUNCTION_ARGS);
Datum       txid_snapshot_out(PG_FUNCTION_ARGS);
Datum       txid_snapshot_recv(PG_FUNCTION_ARGS);
Datum       txid_snapshot_send(PG_FUNCTION_ARGS);

Datum       txid_snapshot_xmin(PG_FUNCTION_ARGS);
Datum       txid_snapshot_xmax(PG_FUNCTION_ARGS);
Datum       txid_snapshot_xip(PG_FUNCTION_ARGS);
Datum       txid_visible_in_snapshot(PG_FUNCTION_ARGS);

Datum       txid_snapshot_active(PG_FUNCTION_ARGS);
Datum       txid_in_snapshot(PG_FUNCTION_ARGS);
Datum       txid_not_in_snapshot(PG_FUNCTION_ARGS);


#endif /* _TXID_H_ */

