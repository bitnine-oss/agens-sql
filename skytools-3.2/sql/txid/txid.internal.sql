-- txid is included in 8.3
