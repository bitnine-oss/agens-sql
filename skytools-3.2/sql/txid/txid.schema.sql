-- ----------
-- txid.sql
--
--	SQL script for loading the transaction ID compatible datatype 
--
--	Copyright (c) 2003-2004, PostgreSQL Global Development Group
--	Author: Jan Wieck, Afilias USA INC.
--
-- ----------

--
-- now the epoch storage
--

CREATE SCHEMA txid;

-- remember txid settings
-- use bigint so we can do arithmetic with it
create table txid.epoch (
	epoch bigint,
	last_value bigint
);

-- make sure there exist exactly one row
insert into txid.epoch values (0, 1);


-- then protect it
create function txid.epoch_guard()
returns trigger as $$
begin
    if TG_OP = 'UPDATE' then
	-- epoch: allow only small increase
	if NEW.epoch > OLD.epoch and NEW.epoch < (OLD.epoch + 3) then
	    return NEW;
	end if;
	-- last_value: allow only increase
	if NEW.epoch = OLD.epoch and NEW.last_value > OLD.last_value then
	    return NEW;
	end if;
    end if;
    raise exception 'bad operation on txid.epoch';
end;
$$ language plpgsql;

-- the trigger
create trigger epoch_guard_trigger
before insert or update or delete on txid.epoch
for each row execute procedure txid.epoch_guard();

