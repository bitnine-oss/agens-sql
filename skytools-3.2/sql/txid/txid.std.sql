-- ----------
-- txid.sql
--
--	SQL script for loading the transaction ID compatible datatype 
--
--	Copyright (c) 2003-2004, PostgreSQL Global Development Group
--	Author: Jan Wieck, Afilias USA INC.
--
-- ----------

set client_min_messages = 'warning';

CREATE DOMAIN txid AS bigint CHECK (value > 0);

--
-- A special transaction snapshot data type for faster visibility checks
--
CREATE OR REPLACE FUNCTION txid_snapshot_in(cstring)
	RETURNS txid_snapshot
	AS 'MODULE_PATHNAME' LANGUAGE C
	IMMUTABLE STRICT;
CREATE OR REPLACE FUNCTION txid_snapshot_out(txid_snapshot)
	RETURNS cstring
	AS 'MODULE_PATHNAME' LANGUAGE C
	IMMUTABLE STRICT;
CREATE OR REPLACE FUNCTION txid_snapshot_recv(internal)
	RETURNS txid_snapshot
	AS 'MODULE_PATHNAME' LANGUAGE C
	IMMUTABLE STRICT;
CREATE OR REPLACE FUNCTION txid_snapshot_send(txid_snapshot)
	RETURNS bytea
	AS 'MODULE_PATHNAME' LANGUAGE C
	IMMUTABLE STRICT;

--
-- The data type itself
--
CREATE TYPE txid_snapshot (
	INPUT = txid_snapshot_in,
	OUTPUT = txid_snapshot_out,
    RECEIVE = txid_snapshot_recv,
    SEND = txid_snapshot_send,
	INTERNALLENGTH = variable,
	STORAGE = extended,
	ALIGNMENT = double
);

--CREATE OR REPLACE FUNCTION get_current_txid()
CREATE OR REPLACE FUNCTION txid_current()
	RETURNS bigint
	AS 'MODULE_PATHNAME', 'txid_current' LANGUAGE C
	STABLE SECURITY DEFINER;

-- CREATE OR REPLACE FUNCTION get_current_snapshot()
CREATE OR REPLACE FUNCTION txid_current_snapshot()
	RETURNS txid_snapshot
	AS 'MODULE_PATHNAME', 'txid_current_snapshot' LANGUAGE C
	STABLE SECURITY DEFINER;

--CREATE OR REPLACE FUNCTION get_snapshot_xmin(txid_snapshot)
CREATE OR REPLACE FUNCTION txid_snapshot_xmin(txid_snapshot)
	RETURNS bigint
	AS 'MODULE_PATHNAME', 'txid_snapshot_xmin' LANGUAGE C
	IMMUTABLE STRICT;

-- CREATE OR REPLACE FUNCTION get_snapshot_xmax(txid_snapshot)
CREATE OR REPLACE FUNCTION txid_snapshot_xmax(txid_snapshot)
	RETURNS bigint
	AS 'MODULE_PATHNAME', 'txid_snapshot_xmax' LANGUAGE C
	IMMUTABLE STRICT;

-- CREATE OR REPLACE FUNCTION get_snapshot_active(txid_snapshot)
CREATE OR REPLACE FUNCTION txid_snapshot_xip(txid_snapshot)
	RETURNS setof bigint
	AS 'MODULE_PATHNAME', 'txid_snapshot_xip' LANGUAGE C
	IMMUTABLE STRICT;


--
-- Special comparision functions used by the remote worker
-- for sync chunk selection
--
CREATE OR REPLACE FUNCTION txid_visible_in_snapshot(bigint, txid_snapshot)
	RETURNS boolean
	AS 'MODULE_PATHNAME', 'txid_visible_in_snapshot' LANGUAGE C
	IMMUTABLE STRICT;
/*
CREATE OR REPLACE FUNCTION txid_in_snapshot(bigint, txid_snapshot)
	RETURNS boolean
	AS 'MODULE_PATHNAME', 'txid_in_snapshot' LANGUAGE C
	IMMUTABLE STRICT;

CREATE OR REPLACE FUNCTION txid_not_in_snapshot(bigint, txid_snapshot)
	RETURNS boolean
	AS 'MODULE_PATHNAME', 'txid_not_in_snapshot' LANGUAGE C
	IMMUTABLE STRICT;
*/
