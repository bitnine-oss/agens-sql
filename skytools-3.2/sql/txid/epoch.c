/*-------------------------------------------------------------------------
 * epoch.c
 *
 *	Detect current epoch.
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <limits.h>

#include "access/transam.h"
#include "executor/spi.h"
#include "miscadmin.h"
#include "catalog/pg_control.h"
#include "access/xlog.h"

#include "txid.h"

/*
 * do a TransactionId -> txid conversion
 */
txid txid_convert_xid(TransactionId xid, TxidEpoch *state)
{
	uint64 epoch;

	/* avoid issues with the the special meaning of 0 */
	if (xid == InvalidTransactionId)
		return MAX_INT64;

	/* return special xid's as-is */
	if (xid < FirstNormalTransactionId)
		return xid;

	/* xid can on both sides on wrap-around */
	epoch = state->epoch;
	if (TransactionIdPrecedes(xid, state->last_value)) {
		if (xid > state->last_value)
			epoch--;
	} else if (TransactionIdFollows(xid, state->last_value)) {
		if (xid < state->last_value)
			epoch++;
	}
	return (epoch << 32) | xid;
}

#if PG_CONTROL_VERSION >= 820

/*
 * PostgreSQl 8.2 keeps track of epoch internally.
 */

void txid_load_epoch(TxidEpoch *state, int try_write)
{
	TransactionId	xid;
	uint32			epoch;

	GetNextXidAndEpoch(&xid, &epoch);

	state->epoch = epoch;
	state->last_value = xid;
}

#else

/*
 * For older PostgreSQL keep epoch in table.
 */

/*
 * this caches the txid_epoch table.
 * The struct should be updated only together with the table.
 */
static TxidEpoch epoch_state = { 0, 0 };

/*
 * load values from txid_epoch table.
 */
static int load_epoch(void)
{
	HeapTuple row;
	TupleDesc rdesc;
	bool isnull = false;
	Datum tmp;
	int res;
	uint64 db_epoch, db_value;

	res = SPI_connect();
	if (res < 0)
		elog(ERROR, "cannot connect to SPI");

	res = SPI_execute("select epoch, last_value from txid.epoch", true, 0);
	if (res != SPI_OK_SELECT)
		elog(ERROR, "load_epoch: select failed?");
	if (SPI_processed != 1)
		elog(ERROR, "load_epoch: there must be exactly 1 row");

	row = SPI_tuptable->vals[0];
	rdesc = SPI_tuptable->tupdesc;

	tmp = SPI_getbinval(row, rdesc, 1, &isnull);
	if (isnull)
		elog(ERROR, "load_epoch: epoch is NULL");
	db_epoch = DatumGetInt64(tmp);

	tmp = SPI_getbinval(row, rdesc, 2, &isnull);
	if (isnull)
		elog(ERROR, "load_epoch: last_value is NULL");
	db_value = DatumGetInt64(tmp);
	
	SPI_finish();

	/*
	 * If the db has lesser values, then some updates were lost.
	 *
	 * Should that be special-cased?  ATM just use db values.
	 * Thus immidiate update.
	 */
	epoch_state.epoch = db_epoch;
	epoch_state.last_value = db_value;
	return 1;
}

/*
 * updates last_value and epoch, if needed
 */
static void save_epoch(void)
{
	int res;
	char qbuf[200];
	uint64 new_epoch, new_value;
	TransactionId xid = GetTopTransactionId();
	TransactionId old_value;

	/* store old state */
	MemoryContext oldcontext = CurrentMemoryContext;
	ResourceOwner oldowner = CurrentResourceOwner;

	/*
	 * avoid changing internal values.
	 */
	new_value = xid;
	new_epoch = epoch_state.epoch;
	old_value = (TransactionId)epoch_state.last_value;
	if (xid < old_value) {
		if (TransactionIdFollows(xid, old_value))
			new_epoch++;
		else
			return;
	}
	sprintf(qbuf, "update txid.epoch set epoch = %llu, last_value = %llu",
				(unsigned long long)new_epoch,
				(unsigned long long)new_value);

	/*
	 * The update may fail in case of SERIALIZABLE transaction.
	 * Try to catch the error and hide it.
	 */
	BeginInternalSubTransaction(NULL);
	PG_TRY();
	{
		/* do the update */
		res = SPI_connect();
		if (res < 0)
			elog(ERROR, "cannot connect to SPI");
		res = SPI_execute(qbuf, false, 0);
		SPI_finish();

		ReleaseCurrentSubTransaction();
	}
	PG_CATCH();
	{
		/* we expect rollback to clean up inner SPI call */
		RollbackAndReleaseCurrentSubTransaction();
		FlushErrorState();
		res = -1;  /* remember failure */
	}
	PG_END_TRY();

	/* restore old state */
	MemoryContextSwitchTo(oldcontext);
	CurrentResourceOwner = oldowner;

	if (res < 0)
		return;

	/*
	 * Seems the update was successful, update internal state too.
	 *
	 * There is a chance that the TX will be rollbacked, but then
	 * another backend will do the update, or this one at next
	 * checkpoint.
	 */
	epoch_state.epoch = new_epoch;
	epoch_state.last_value = new_value;
}

static void check_epoch(int update_prio)
{
	TransactionId xid = GetTopTransactionId();
	TransactionId recheck, tx_next;
	int ok = 1;

	/* should not happen, but just in case */
	if (xid == InvalidTransactionId)
		return;

	/* new backend */
	if (epoch_state.last_value == 0)
		load_epoch();
	
	/* try to avoid concurrent access */
	if (update_prio)
		recheck = 50000 + 100 * (MyProcPid & 0x1FF);
	else
		recheck = 300000 + 1000 * (MyProcPid & 0x1FF);

	/* read table */
	tx_next = (TransactionId)epoch_state.last_value + recheck;
	if (TransactionIdFollows(xid, tx_next))
		ok = load_epoch();

	/*
	 * check if save is needed.  last_value may be updated above.
	 */
	tx_next = (TransactionId)epoch_state.last_value + recheck;
	if (!ok || TransactionIdFollows(xid, tx_next))
		save_epoch();
}

void txid_load_epoch(TxidEpoch *state, int try_write)
{
	check_epoch(try_write);

	state->epoch = epoch_state.epoch;
	state->last_value = epoch_state.last_value;
}


#endif
