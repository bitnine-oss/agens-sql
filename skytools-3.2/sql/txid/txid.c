/*-------------------------------------------------------------------------
 * txid.c
 *
 *	Safe handling of transaction ID's.
 *
 *	Copyright (c) 2003-2004, PostgreSQL Global Development Group
 *	Author: Jan Wieck, Afilias USA INC.
 *
 *	64-bit output: Marko Kreen, Skype Technologies
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <limits.h>

#include "access/xact.h"
#include "funcapi.h"
#include "lib/stringinfo.h"
#include "libpq/pqformat.h"

#include "txid.h"

#ifdef INT64_IS_BUSTED
#error txid needs working int64
#endif

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

#ifndef SET_VARSIZE
#define SET_VARSIZE(x, len) VARATT_SIZEP(x) = len
#endif

/* txid will be signed int8 in database, so must limit to 63 bits */
#define MAX_TXID   UINT64CONST(0x7FFFFFFFFFFFFFFF)

/*
 * If defined, use bsearch() function for searching
 * txid's inside snapshots that have more than given values.
 */
#define USE_BSEARCH_FOR 100


/*
 * public functions
 */

PG_FUNCTION_INFO_V1(txid_current);
PG_FUNCTION_INFO_V1(txid_snapshot_in);
PG_FUNCTION_INFO_V1(txid_snapshot_out);
PG_FUNCTION_INFO_V1(txid_snapshot_recv);
PG_FUNCTION_INFO_V1(txid_snapshot_send);
PG_FUNCTION_INFO_V1(txid_current_snapshot);
PG_FUNCTION_INFO_V1(txid_snapshot_xmin);
PG_FUNCTION_INFO_V1(txid_snapshot_xmax);

/* new API in 8.3 */
PG_FUNCTION_INFO_V1(txid_visible_in_snapshot);
PG_FUNCTION_INFO_V1(txid_snapshot_xip);

/* old API */
PG_FUNCTION_INFO_V1(txid_in_snapshot);
PG_FUNCTION_INFO_V1(txid_not_in_snapshot);
PG_FUNCTION_INFO_V1(txid_snapshot_active);

/*
 * utility functions
 */

static int _cmp_txid(const void *aa, const void *bb)
{
	const uint64 *a = aa;
	const uint64 *b = bb;
	if (*a < *b)
		return -1;
	if (*a > *b)
		return 1;
	return 0;
}

static void sort_snapshot(TxidSnapshot *snap)
{
	qsort(snap->xip, snap->nxip, sizeof(txid), _cmp_txid);
}

static StringInfo
buf_init(txid xmin, txid xmax)
{
	TxidSnapshot snap;
	StringInfo buf;

	snap.xmin = xmin;
	snap.xmax = xmax;
	snap.nxip = 0;

	buf = makeStringInfo();
	appendBinaryStringInfo(buf, (char *)&snap, offsetof(TxidSnapshot, xip));
	return buf;
}

static void
buf_add_txid(StringInfo buf, txid xid)
{
	TxidSnapshot *snap = (TxidSnapshot *)buf->data;
	snap->nxip++;
	appendBinaryStringInfo(buf, (char *)&xid, sizeof(xid));
}

static TxidSnapshot *
buf_finalize(StringInfo buf)
{
	TxidSnapshot *snap = (TxidSnapshot *)buf->data;
	SET_VARSIZE(snap, buf->len);

	/* buf is not needed anymore */
	buf->data = NULL;
	pfree(buf);

	return snap;
}

static TxidSnapshot *
parse_snapshot(const char *str)
{
	txid		xmin;
	txid		xmax;
	txid		last_val = 0, val;
	char	   *endp;
	StringInfo  buf;

	xmin = (txid) strtoull(str, &endp, 0);
	if (*endp != ':')
		goto bad_format;
	str = endp + 1;

	xmax = (txid) strtoull(str, &endp, 0);
	if (*endp != ':')
		goto bad_format;
	str = endp + 1;

	/* it should look sane */
	if (xmin >= xmax || xmin == 0 || xmax > MAX_INT64)
		goto bad_format;

	/* allocate buffer */
	buf = buf_init(xmin, xmax);

	/* loop over values */
	while (*str != '\0')
	{
		/* read next value */
		val = (txid) strtoull(str, &endp, 0);
		str = endp;

		/* require the input to be in order */
		if (val < xmin || val <= last_val || val >= xmax)
			goto bad_format;
		
		buf_add_txid(buf, val);
		last_val = val;

		if (*str == ',')
			str++;
		else if (*str != '\0')
			goto bad_format;
	}

	return buf_finalize(buf);

bad_format:
	elog(ERROR, "illegal txid_snapshot input format");
	return NULL;
}

/*
 * Public functions
 */

/*
 *		txid_current	- Return the current transaction ID as txid
 */
Datum
txid_current(PG_FUNCTION_ARGS)
{
	txid val;
	TxidEpoch state;

	txid_load_epoch(&state, 0);

	val = txid_convert_xid(GetTopTransactionId(), &state);

	PG_RETURN_INT64(val);
}

/*
 * txid_current_snapshot	-	return current snapshot
 */
Datum
txid_current_snapshot(PG_FUNCTION_ARGS)
{
	TxidSnapshot *snap;
	unsigned num, i, size;
	TxidEpoch state;

	if (SerializableSnapshot == NULL)
		elog(ERROR, "get_current_snapshot: SerializableSnapshot == NULL");

	txid_load_epoch(&state, 1);

	num = SerializableSnapshot->xcnt;
	size = offsetof(TxidSnapshot, xip) + sizeof(txid) * num;
	snap = palloc(size);
	SET_VARSIZE(snap, size);
	snap->xmin = txid_convert_xid(SerializableSnapshot->xmin, &state);
	snap->xmax = txid_convert_xid(SerializableSnapshot->xmax, &state);
	snap->nxip = num;
	for (i = 0; i < num; i++)
		snap->xip[i] = txid_convert_xid(SerializableSnapshot->xip[i], &state);

	/* we want them guaranteed ascending order */
	sort_snapshot(snap);

	PG_RETURN_POINTER(snap);
}

/*
 *		txid_snapshot_in	- input function for type txid_snapshot
 */
Datum
txid_snapshot_in(PG_FUNCTION_ARGS)
{
	TxidSnapshot *snap;
	char	   *str = PG_GETARG_CSTRING(0);

	snap = parse_snapshot(str);
	PG_RETURN_POINTER(snap);
}

/*
 *		txid_snapshot_out	- output function for type txid_snapshot
 */
Datum
txid_snapshot_out(PG_FUNCTION_ARGS)
{
	TxidSnapshot   *snap;
	StringInfoData	str;
	int				i;

	snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(0);

	initStringInfo(&str);

	appendStringInfo(&str, "%llu:", (unsigned long long)snap->xmin);
	appendStringInfo(&str, "%llu:", (unsigned long long)snap->xmax);

	for (i = 0; i < snap->nxip; i++)
	{
		appendStringInfo(&str, "%s%llu", ((i > 0) ? "," : ""),
						 (unsigned long long)snap->xip[i]);
	}

	PG_FREE_IF_COPY(snap, 0);

	PG_RETURN_CSTRING(str.data);
}

/*
 * txid_snapshot_recv(internal) returns txid_snapshot
 *
 *		binary input function for type txid_snapshot
 *
 *		format: int4 nxip, int8 xmin, int8 xmax, int8 xip
 */
Datum
txid_snapshot_recv(PG_FUNCTION_ARGS)
{
	StringInfo  buf = (StringInfo) PG_GETARG_POINTER(0);
	TxidSnapshot *snap;
	txid last = 0;
	int nxip;
	int i;
	int avail;
	int expect;
	txid xmin, xmax;

	/*
	 * load nxip and check for nonsense.
	 *
	 * (nxip > avail) check is against int overflows in 'expect'.
	 */
	nxip = pq_getmsgint(buf, 4);
	avail = buf->len - buf->cursor;
	expect = 8 + 8 + nxip * 8;
	if (nxip < 0 || nxip > avail || expect > avail)
		goto bad_format;

	xmin = pq_getmsgint64(buf);
	xmax = pq_getmsgint64(buf);
	if (xmin == 0 || xmax == 0 || xmin > xmax || xmax > MAX_TXID)
		goto bad_format;

	snap = palloc(TXID_SNAPSHOT_SIZE(nxip));
	snap->xmin = xmin;
	snap->xmax = xmax;
	snap->nxip = nxip;
	SET_VARSIZE(snap, TXID_SNAPSHOT_SIZE(nxip));

	for (i = 0; i < nxip; i++)
	{
		txid cur =  pq_getmsgint64(buf);
		if (cur <= last || cur < xmin || cur >= xmax)
			goto bad_format;
		snap->xip[i] = cur;
		last = cur;
	}
	PG_RETURN_POINTER(snap);

bad_format:
	elog(ERROR, "invalid snapshot data");
	return (Datum)NULL;
}

/*
 * txid_snapshot_send(txid_snapshot) returns bytea
 *
 *		binary output function for type txid_snapshot
 *
 *		format: int4 nxip, int8 xmin, int8 xmax, int8 xip
 */
Datum
txid_snapshot_send(PG_FUNCTION_ARGS)
{
	TxidSnapshot *snap = (TxidSnapshot *)PG_GETARG_VARLENA_P(0);
	StringInfoData buf;
	uint32 i;

	pq_begintypsend(&buf);
	pq_sendint(&buf, snap->nxip, 4);
	pq_sendint64(&buf, snap->xmin);
	pq_sendint64(&buf, snap->xmax);
	for (i = 0; i < snap->nxip; i++)
		pq_sendint64(&buf, snap->xip[i]);
	PG_RETURN_BYTEA_P(pq_endtypsend(&buf));
}


static int
_txid_in_snapshot(txid value, const TxidSnapshot *snap)
{
	if (value < snap->xmin)
		return true;
	else if (value >= snap->xmax)
		return false;
#ifdef USE_BSEARCH_FOR
	else if (snap->nxip >= USE_BSEARCH_FOR)
	{
		void *res;
		res = bsearch(&value, snap->xip, snap->nxip, sizeof(txid), _cmp_txid);
		return (res) ? false : true;
	}
#endif
	else
	{
		int			i;
		for (i = 0; i < snap->nxip; i++)
		{
			if (value == snap->xip[i])
				return false;
		}
		return true;
	}
}

/*
 * txid_in_snapshot	- is txid visible in snapshot ?
 */
Datum
txid_in_snapshot(PG_FUNCTION_ARGS)
{
	txid value = PG_GETARG_INT64(0);
	TxidSnapshot *snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(1);
	int			res;
	
	res = _txid_in_snapshot(value, snap) ? true : false;

	PG_FREE_IF_COPY(snap, 1);
	PG_RETURN_BOOL(res);
}

/*
 * changed api
 */
Datum
txid_visible_in_snapshot(PG_FUNCTION_ARGS)
{
	txid value = PG_GETARG_INT64(0);
	TxidSnapshot *snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(1);
	int			res;
	
	res = _txid_in_snapshot(value, snap) ? true : false;

	PG_FREE_IF_COPY(snap, 1);
	PG_RETURN_BOOL(res);
}


/*
 * txid_not_in_snapshot	- is txid invisible in snapshot ?
 */
Datum
txid_not_in_snapshot(PG_FUNCTION_ARGS)
{
	txid		value = PG_GETARG_INT64(0);
	TxidSnapshot *snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(1);
	int			res;

	res = _txid_in_snapshot(value, snap) ? false : true;

	PG_FREE_IF_COPY(snap, 1);
	PG_RETURN_BOOL(res);
}

/*
 * txid_snapshot_xmin	-	return snapshot's xmin
 */
Datum
txid_snapshot_xmin(PG_FUNCTION_ARGS)
{
	TxidSnapshot *snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(0);
	txid res = snap->xmin;
	PG_FREE_IF_COPY(snap, 0);
	PG_RETURN_INT64(res);
}

/*
 * txid_snapshot_xmin	-	return snapshot's xmax
 */
Datum
txid_snapshot_xmax(PG_FUNCTION_ARGS)
{
	TxidSnapshot *snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(0);
	txid res = snap->xmax;
	PG_FREE_IF_COPY(snap, 0);
	PG_RETURN_INT64(res);
}

/* remember state between function calls */
struct snap_state {
	int pos;
	TxidSnapshot *snap;
};

/*
 * txid_snapshot_active		- returns uncommitted TXID's in snapshot.
 */
Datum
txid_snapshot_xip(PG_FUNCTION_ARGS)
{
	FuncCallContext *fctx;
	struct snap_state *state;

	if (SRF_IS_FIRSTCALL()) {
		TxidSnapshot *snap;
		int statelen;

		snap = (TxidSnapshot *) PG_GETARG_VARLENA_P(0);
		
		fctx = SRF_FIRSTCALL_INIT();
		statelen = sizeof(*state) + VARSIZE(snap);
		state = MemoryContextAlloc(fctx->multi_call_memory_ctx, statelen);
		state->pos = 0;
		state->snap = (TxidSnapshot *)((char *)state + sizeof(*state));
		memcpy(state->snap, snap, VARSIZE(snap));
		fctx->user_fctx = state;

		PG_FREE_IF_COPY(snap, 0);
	}
	fctx = SRF_PERCALL_SETUP();
	state = fctx->user_fctx;
	if (state->pos < state->snap->nxip) {
		Datum res = Int64GetDatum(state->snap->xip[state->pos]);
		state->pos++;
		SRF_RETURN_NEXT(fctx, res);
	} else {
		SRF_RETURN_DONE(fctx);
	}
}

/* old api */
Datum
txid_snapshot_active(PG_FUNCTION_ARGS)
{
	return txid_snapshot_xip(fcinfo);
}

