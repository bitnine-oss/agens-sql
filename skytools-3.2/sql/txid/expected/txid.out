-- init
\set ECHO none
-- i/o
select '12:13:'::txid_snapshot;
 txid_snapshot 
---------------
 12:13:
(1 row)

select '12:13:1,2'::txid_snapshot;
ERROR:  illegal txid_snapshot input format
-- errors
select '31:12:'::txid_snapshot;
ERROR:  illegal txid_snapshot input format
select '0:1:'::txid_snapshot;
ERROR:  illegal txid_snapshot input format
select '12:13:0'::txid_snapshot;
ERROR:  illegal txid_snapshot input format
select '12:13:2,1'::txid_snapshot;
ERROR:  illegal txid_snapshot input format
create table snapshot_test (
	nr	integer,
	snap	txid_snapshot
);
insert into snapshot_test values (1, '12:13:');
insert into snapshot_test values (2, '12:20:13,15,18');
insert into snapshot_test values (3, '100001:100009:100005,100007,100008');
select snap from snapshot_test order by nr;
                snap                
------------------------------------
 12:13:
 12:20:13,15,18
 100001:100009:100005,100007,100008
(3 rows)

select  txid_snapshot_xmin(snap),
	txid_snapshot_xmax(snap),
	txid_snapshot_xip(snap)
from snapshot_test order by nr;
 txid_snapshot_xmin | txid_snapshot_xmax | txid_snapshot_xip 
--------------------+--------------------+-------------------
                 12 |                 20 |                13
                 12 |                 20 |                15
                 12 |                 20 |                18
             100001 |             100009 |            100005
             100001 |             100009 |            100007
             100001 |             100009 |            100008
(6 rows)

select id, txid_visible_in_snapshot(id, snap)
from snapshot_test, generate_series(11, 21) id
where nr = 2;
 id | txid_visible_in_snapshot 
----+--------------------------
 11 | t
 12 | t
 13 | f
 14 | t
 15 | f
 16 | t
 17 | t
 18 | f
 19 | t
 20 | f
 21 | f
(11 rows)

-- test current values also
select txid_current() >= txid_snapshot_xmin(txid_current_snapshot());
 ?column? 
----------
 t
(1 row)

-- select txid_current_txid() < txid_snapshot_xmax(txid_current_snapshot());
-- select txid_in_snapshot(txid_current_txid(), txid_current_snapshot()),
--    txid_not_in_snapshot(txid_current_txid(), txid_current_snapshot());
